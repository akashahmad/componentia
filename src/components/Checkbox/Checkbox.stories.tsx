import * as React from "react";

import type { Meta, StoryObj } from "@storybook/react";

import { Checkbox } from "./Checkbox";

const meta: Meta<typeof Checkbox> = {
    component: Checkbox,
    tags: ["autodocs"],
};

export default meta;
type Story = StoryObj<typeof Checkbox>;

export const Default: Story = {
    args: {
        label: "Checkbox Text",
    },
    argTypes: {
        label: {
            table: {
                type: { summary: "string" },
                defaultValue: { summary: "undefined" },
            },
        },
        value: {
            table: {
                defaultValue: { summary: '"" ' },
            },
        },
        onChange: {
            table: {
                defaultValue: { summary: "undefined" },
            },
        },
    },
    render: props => {
        return <Checkbox {...props} />;
    },
};

export const ExampleUsage: Story = {
    args: {
        label: "Checkbox Text",
    },
    argTypes: {},
    render: () => {
        const [value, setValue] = React.useState("xyz");

        return (
            <div>
                <Checkbox
                    value={"abc"}
                    onChange={value => setValue(value)}
                    checked={value === "abc"}
                    label="Checkbox Type (value: abc)"
                />
                <Checkbox
                    value={"yz"}
                    onChange={value => setValue(value)}
                    checked={value === "xyz"}
                    label="Checkmark Type (value: xyz)"
                    boxType="checkmark"
                />
                Current Value: {value}
            </div>
        );
    },
};
