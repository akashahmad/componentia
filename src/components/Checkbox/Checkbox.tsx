import * as React from "react";

import cx from "classnames";

import { Typography } from "../Typography/Typography";
import { UncheckedIcon } from "../icons/UncheckedIcon";
import { RectangularChecked } from "../icons/RectangularChecked";
import { CheckmarkedChecked } from "../icons/CheckmarkChecked";

export function Checkbox({ label, checked = false, boxType = "default", value, onChange, ...other }: CheckboxProps) {
    const [isChecked, setIsChecked] = React.useState(false);

    React.useEffect(() => {
        setIsChecked(checked);
    }, [checked]);

    const handleChecked = () => {
        setIsChecked(prev => !prev);

        onChange(value);
    };

    const checkedIcon = boxType === "default" ? <RectangularChecked /> : <CheckmarkedChecked />;

    return (
        <div className={cx("flex items-center select-none cursor-pointer w-max")} onClick={handleChecked}>
            <span className={cx("inline-block mr-[10px] cursor-pointer")}>
                {isChecked ? checkedIcon : <UncheckedIcon />}
            </span>
            {label && <Typography variant="body2">{label}</Typography>}
            <input
                checked={isChecked}
                onChange={e => onChange(e.target.value)}
                value={value}
                type="checkbox"
                className={cx("appearance-none opacity-0")}
                {...other}
            />
        </div>
    );
}

type CheckboxProps = CheckboxCustomProps;
interface CheckboxCustomProps {
    label?: React.ReactNode;
    checked?: boolean;
    value: string;
    onChange: (_value: string) => void;
    boxType?: "default" | "checkmark";
}
