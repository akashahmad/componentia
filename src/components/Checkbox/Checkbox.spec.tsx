import * as React from "react";

import { render, screen } from "@testing-library/react";
import { Checkbox } from "./Checkbox";

test("Checkbox component is rendering properly", async () => {
    const TEXT = "TEXT";

    const { container } = render(<Checkbox value={TEXT} onChange={() => {}} checked label={TEXT} />);

    expect(await screen.getByText(TEXT)).not.toBeNull();
    expect(container).toMatchSnapshot();
});
