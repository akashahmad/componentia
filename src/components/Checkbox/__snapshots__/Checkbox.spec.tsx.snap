// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component is rendering properly 1`] = `
<div>
  <div
    class="flex items-center select-none cursor-pointer w-max"
  >
    <span
      class="inline-block mr-[10px] cursor-pointer"
    >
      <svg
        fill="none"
        height="20"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="#F1F4FF"
          height="19"
          rx="3.5"
          stroke="#009AEE"
          width="19"
          x="0.5"
          y="0.5"
        />
        <rect
          fill="#009AEE"
          height="10"
          rx="2"
          width="10"
          x="5"
          y="5"
        />
      </svg>
    </span>
    <p
      class="text-base leading-6 font-normal"
    >
      TEXT
    </p>
    <input
      class="appearance-none opacity-0"
      type="checkbox"
      value="TEXT"
    />
  </div>
</div>
`;
